### PR TITLE
Initial Cinematic Dialog System

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
### tl;dr:
Major new file is DialogSystem.js.
Other changes are minor housekeeping changes and improvements.

## Simplified data transformation to get text on the screen.

DialogSystem

## Features:

  - `/cinematic` route now redirects to `/` if not an admin.
  - Positioning text bubble
  - A simple template language leveraging a [super light library](https://blueimp.github.io/JavaScript-Templates/) that we've [forked and simplified](https://github.com/codecombat/JavaScript-Templates/tree/remove-load).
  - Localization added to schema and dialog bubbles correctly handle language.
  - Left and Right svg speaker bubbles. (svg programmer art)
  - Some light tests on the most important pieces. (text transpilation and transformation)
  - Clear all current text bubbles command.
  - All systems are now iterated for `parseDialogNode` per dialogNode.
  - All systems are now iterated for `parseSetupShot` per setupShot node.
  - i18n changes added to cinematic schema.
  - moved some logic into more rational places. Commands are now in a single file, and systems handle creating their own commands.

### CODE REVIEW

File: app/views/play/cinematic/Command/AbstractCommand.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// Throws and error if `import ... from ..` syntax.
```
Comment (AndrewJakubowicz) [inline app/views/play/cinematic/Command/AbstractCommand.js]:
```suggestion
// Throws an error if `import ... from ..` syntax.
```

File: app/views/play/cinematic/CinematicLankBoss.js [hunk 2]
```diff
@@ -1,28 +1,63 @@
...
     }
+    this.loader = loader
+  }
+
+  /**
+   * Returns a list of commands that correctly set up the shot.
+   * @param {Shot} shot - the cinematic shot data.
+   */
+  parseSetupShot (shot) {
+    const leftCharSlug = getLeftCharacterThangTypeSlug(shot)
```
Comment (AndrewJakubowicz) [inline app/views/play/cinematic/CinematicLankBoss.js]: The code in this method was a move with no logic changes.

File: app/views/play/cinematic/CinematicLankBoss.js [hunk 3]
```diff
@@ -128,40 +165,6 @@ export default class CinematicLankBoss {
   }
 }
 
-/**
- * AnimeCommand is used to turn animejs animation tweens into commands that the Command Runner can play and cancel.
- */
-class AnimeCommand extends AbstractCommand {
```
Comment (AndrewJakubowicz) [inline app/views/play/cinematic/CinematicLankBoss.js]: This code was also moved with no logic changes.

File: app/views/play/cinematic/Command/AbstractCommand.js [hunk 4]
```diff
@@ -103,3 +104,54 @@ export class Noop extends AbstractCommand {
     return Promise.resolve()
   }
 }
+
+/**
+ * AnimeCommand is used to turn animejs animation tweens into commands that the Command Runner can play and cancel.
+ */
+export class AnimeCommand extends AbstractCommand {
```
Comment (AndrewJakubowicz) [inline app/views/play/cinematic/Command/AbstractCommand.js]: `AnimeCommand` was moved here with no logic changes.

File: app/views/play/cinematic/Command/CinematicParser.js [hunk 5]
```diff
@@ -1,8 +1,40 @@
-import { getLeftCharacterThangTypeSlug, getRightCharacterThangTypeSlug } from '../../../../schemas/selectors/cinematic'
+import AbstractCommand from './AbstractCommand'
```
Comment (AndrewJakubowicz) [inline app/views/play/cinematic/Command/CinematicParser.js]: This file has logic changes.
We iterate through systems checking if they have particular setup methods.
If they do we give them cinematic data and concatenate a list of commands returned from all systems.

File: app/views/CinematicCanvas.vue [hunk 6]
```diff
@@ -28,6 +28,10 @@ export default {
     enterDisabled: true
   }),
   mounted: function() {
+    if (!me.isAdmin()) {
```
Comment (shubhi1092) [inline app/views/CinematicCanvas.vue]: Alternatively, we can create a feature flag for this and do the admin check there like this: https://github.com/codecombat/codecombat/blob/master/app/models/User.coffee#L584

File: app/views/play/cinematic/CinematicLankBoss.js [hunk 7]
```diff
@@ -1,28 +1,63 @@
 import anime from 'animejs/lib/anime.es.js'
-import Promise from 'bluebird'
-import AbstractCommand, { Noop } from './Command/AbstractCommand'
+import { Noop, AnimeCommand } from './Command/AbstractCommand'
+import { getLeftCharacterThangTypeSlug, getRightCharacterThangTypeSlug } from '../../../schemas/selectors/cinematic'
 
 const Lank = require('lib/surface/Lank')
 
-Promise.config({
```
Comment (shubhi1092) [inline app/views/play/cinematic/CinematicLankBoss.js]: Why are we removing this

File: app/views/play/cinematic/CinematicLankBoss.js [hunk 8]
```diff
@@ -68,8 +103,10 @@ export default class CinematicLankBoss {
       y: pos.y,
       duration: ms,
       autoplay: false,
-      // Update required to ensure Lank sprite position is moved.
-      update: () => { this[side].thang.stateChanged = true }
+      easing: 'easeInOutQuart',
```
Comment (shubhi1092) [inline app/views/play/cinematic/CinematicLankBoss.js]: What is this for?

File: app/views/play/cinematic/Command/AbstractCommand.js (hunk 4 above)
Comment (shubhi1092) [inline app/views/play/cinematic/Command/AbstractCommand.js]: Why did we move this from LankBoss to here?

File: app/views/play/cinematic/Command/CinematicParser.js [hunk 9]
```diff
@@ -13,33 +45,57 @@ export const parseShot = (shot, systems) => {
...
+ * @param {*} systems 
  */
-const parseSetup = (shot, systems) => {
-  const { cinematicLankBoss, loader } = systems
-  const setupCommands = []
+const parseDialogNode = (dialogNode, systems) =>
+  Object.values(systems)
+    .filter(sys => sys !== undefined && typeof sys.parseDialogNode === 'function')
+    .reduce((commands, sys) => (
+      [...commands, ...sys.parseDialogNode(dialogNode)]
```
Comment (shubhi1092) [inline app/views/play/cinematic/Command/CinematicParser.js]: Do we not need to check if `sys.parseDialogNode(dialogNode)` returns an array similar to `parseSetup` above?

File: app/views/play/cinematic/cinematicController.js [hunk 10]
```diff
@@ -53,14 +54,25 @@ export class CinematicController {
...
     this.systems.cinematicLankBoss = new CinematicLankBoss({
       groundLayer: this.stubRequiredLayer,
       layerAdapter: this.layerAdapter,
-      camera: this.systems.camera
+      camera: this.systems.camera,
+      loader: this.systems.loader
     })
 
-    this.systems.loader = new Loader({ slug })
+    this.commands = []
```
Comment (shubhi1092) [inline app/views/play/cinematic/cinematicController.js]: Duplicate of line 77 below

File: app/views/play/cinematic/Command/CinematicParser.js (hunk 9 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/CinematicParser.js]: Great comment. I had apparently not done this in this PR but include it in the next PR.
Thanks for this excellent pickup.

File: app/views/play/cinematic/Command/AbstractCommand.js (hunk 4 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/AbstractCommand.js]: Because we're able to reuse this Command for the dialogSystem.

File: app/views/play/cinematic/CinematicLankBoss.js (hunk 8 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/CinematicLankBoss.js]: The default animation ease function has bounce which makes the characters bounce around a little bit too much. This is a much more pleasing easing function.

File: app/views/play/cinematic/CinematicLankBoss.js (hunk 7 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/CinematicLankBoss.js]: We don't need the cancellable promise, as it was being used in AnimeCommand.

File: app/views/CinematicCanvas.vue (hunk 6 above)
Comment (AndrewJakubowicz) [reply app/views/CinematicCanvas.vue]: That's a brilliant idea.

File: app/views/play/cinematic/cinematicController.js (hunk 10 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/cinematicController.js]: Good catch

File: app/views/play/cinematic/Command/CinematicParser.js [hunk 11]
```diff
@@ -13,33 +45,57 @@ export const parseShot = (shot, systems) => {
...
+ * Returns a 2d array of command nodes. Each inner array is passed into the
+ * CommandRunner and run until user interuption or conclusion. Then between the
+ * inner arrays the cinematicController waits before running the next command batch.
+ *
+ * @param {Shot} shot The shot that is being run.
+ * @param {Object} systems The systems.
+ * @returns {AbstractCommand[][]} A 2d array of commands.
+ */
+export const parseShot = (shot, systems) => {
+  const setupCommands = parseSetup(shot, systems) || []
```
Comment (shubhi1092) [inline app/views/play/cinematic/Command/CinematicParser.js]: Instead of sending all the systems to `parseSetup` and `parseDialogNode`, and then filtering there, I think it would be cleaner to filter here and then send only the specific system to `parseSetup` and `parseDialogNode`

Comment (shubhi1092) [reply app/views/play/cinematic/Command/CinematicParser.js]: Can also probably rename `systems` to `cinematicSystems`?

File: app/views/play/cinematic/DialogSystem.js [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+    const bbox = textDiv.getBoundingClientRect()
+    const width = (bbox.right - bbox.left) + 2 * padding
+    const height = (bbox.bottom - bbox.top) + 2 * padding
+
+    const svgGroup = createSvgShape({
+      x,
+      y: y - height,
+      width,
+      height,
+      clazz: this.id,
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: Should it be `class` ?

File: app/views/play/cinematic/DialogSystem.js [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+ * @property {'left'|'right'} side
+ * @property {string} class - the unique class for this bubble
+ */
+
+/**
+ * Returns a svg speech bubble that can be attached to an svg canvas.
+ * @param {SvgBubbleOptions} svgOptions - Svg options for generating the svg shape.
+ * @returns {SVGGElement} the group element of the svg bubble.
+ */
+function createSvgShape ({ x, y, width, height, side, clazz }) {
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: `clazz` -> `class`

File: app/views/play/cinematic/DialogSystem.js (hunk 12 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: Yeah this was some bad naming. `class` is a reserved keyword so I can't use that. `clazz` is not an easy name. It's flat out confusing. Great pickup.

File: app/views/play/cinematic/DialogSystem.js [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+        result += ' '
+      }
+      return result
+    })
+  }
+
+  return wrapWord(htmlString, / ?([^ ])+ ?/)
+}
+
+// Polyfill for `node.remove` method.
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: Can this polyfill be moved to a common place where app is initialized?

File: app/views/play/cinematic/DialogSystem.js [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+   * E.g. a context object of: `{ name: 'Mary' }` can then be used in the following
+   * dialog text:
+   *   `Hello, <%=o.name%>`
+   * Which then appears as:
+   *   `Hello, Mary`
+   *
+   * @param {Object} context - used for interpolation.
+   */
+  set templateContext (context) {
+    this._context = context
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: Why are we not setting this in the constructor? Also is it generally called template context in the language of text interpolation, otherwise the name `context` doesnt look intuitive to me.

File: app/views/play/cinematic/DialogSystem.js [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+ * This lets us calculate the bounding box of the html element.
+ * To prevent the possible occurrance of line breaks in the middle of a word
+ * we must also wrap words with a span tag.
+ *
+ * @param {string} htmlString is a raw string representation of html
+ * @param {function} wrapLetterString takes a letter and can return a wrapped letter.
+ * @param {function} wrapWordString takes a word and can return a wrapped word.
+ * @returns {string} transformed html string.
+ */
+export function wrapText (htmlString, wrapLetterString, wrapWordString) {
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: Why are we exporting processText and wrapText, can we move them to a common place like `utils` if we want them to be re-usable?

File: test/app/lib/cinematic/DialogSystem.spec.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
+/* eslint-env jasmine */
+import { processText, wrapText } from '../../../../app/views/play/cinematic/DialogSystem'
+
+const dialogNode = text => ({ text })
+
+describe('dialogSystem', () => {
+  describe('text processing', () => {
```
Comment (shubhi1092) [inline test/app/lib/cinematic/DialogSystem.spec.js]: it looks like this is only for testing wrapText, can modify description here to specify that

File: app/views/play/cinematic/DialogSystem.js [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+  set templateContext (context) {
+    this._context = context
+  }
+
+  /**
+   * The system method that is run on every dialogNode.
+   * @param {import('../../../schemas/selectors/cinematic').DialogNode} dialogNode
+   * @returns {AbstractCommand[]}
+   */
+  parseDialogNode (dialogNode) {
```
Comment (shubhi1092) [inline app/views/play/cinematic/DialogSystem.js]: Looks like we can also add test coverage for this method.

File: app/views/play/cinematic/DialogSystem.js (hunk 12 above)
Comment (shubhi1092) [reply app/views/play/cinematic/DialogSystem.js]: Oh i see, can we use `className` in that case.

**Comment by shubhi1092** [comment]
Added some questions and suggestions. Otherwise looks good!

**Comment by shubhi1092** [comment]
I generally feel that the dialogSystem file is too big, and can probably be refactored.

File: app/views/play/cinematic/DialogSystem.js (hunk 16 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: I agree and really don't like these exports, so I'll give my problem hoping for a solution.

I wouldn't want to move these files to `utils` as they're very specific to the Dialog System. Specifically the `wrapText` function is very specific wrapping text in individual tags.
However these are also not easy functions so I want to ensure they have some test coverage.

A test that tests public behavior of the DialogSystem is difficult due to testing side effects. 
Thus I'm exporting these for testing to ensure their behavior is predictable for the system depending on them.

Interested in other tactics to solve the above problem.

Comment (shubhi1092) [reply app/views/play/cinematic/DialogSystem.js]: How about a common file thats specific to dialog system like `dialogSystemHelper`

File: app/views/play/cinematic/DialogSystem.js (hunk 15 above)
Comment (shubhi1092) [reply app/views/play/cinematic/DialogSystem.js]: I like `templateData`, or maybe `templateDataParameters` since parameter is the term used in their documentation.

File: app/views/play/cinematic/DialogSystem.js (hunk 16 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: hmmmm. The more I think about that the more I like it.
I see keeping the functions in `dialogSystem.js` as one option as they're very related.
But then they are exported muddying up with public exports of `dialogSystem`. Whilst placing them in a `dialogSystemHelper` file would allow them to be exported but it makes their intent clearer.
I'm a fan of the `dialogSystemHelper` idea 🚀 👏

File: app/views/play/cinematic/Command/CinematicParser.js (hunk 11 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/CinematicParser.js]: For now I will not change  the filtering (as I think I might change this a little more in the next PR). But I did make  the name more clear. Thanks!

File: app/views/play/cinematic/DialogSystem.js (hunk 12 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: That's such a better name.

File: app/views/play/cinematic/DialogSystem.js (hunk 14 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: This is an amazing comment. Moving it below another polyfill in the `vendor.js` file.

File: app/views/play/cinematic/DialogSystem.js (hunk 16 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: This is also excellent because the `dialogsystem.js` file currently contains 2 ideas. The idea of drawing html nodes on a div canvas and the idea of drawing svgs onto an svg canvas.

These could both be split into separate files and the DialogSystem just uses them.
Not doing the full refactor in  this PR. But creating the initial folder and moving out the functions into the dialogSystemHelper file.

File: app/views/play/cinematic/DialogSystem.js (hunk 18 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/DialogSystem.js]: I agree but am still holding off as this method changes slightly again in the next  PR.
Your feedback is noted.

**Comment by AndrewJakubowicz** [comment]
@shubhi1092 I addressed your concerns. Some of them are addressed in the next PR with other changes.
Generally made the DialogSystem file smaller, which will encourage the files to stay smaller moving forward. That was an excellent suggestion.
